File: src/checks/webhooks-crud.ts
import { Check, CheckContext, CheckResult } from '../types.js';
import { createClient, DripError } from '../drip-client.js';

// Extended context type to store webhook ID across checks
type WebhookContext = CheckContext & { webhookId?: string };

export const webhookCreateCheck: Check = {
  name: 'Webhook Create',
  description: 'Create webhook endpoint',
  async run(ctx: CheckContext): Promise<CheckResult> {
    const start = performance.now();
    const client = createClient(ctx);

    try {
      // Access the SDK directly
      const sdk = (client as unknown as {
        sdk: {
          createWebhook?: (data: { url: string; events: string[] }) => Promise<{ id: string; url: string; events: string[] }>;
        };
      }).sdk;

      if (!sdk.createWebhook) {
        const duration = performance.now() - start;
        return {
          name: 'Webhook Create',
          success: true,
          duration,
          message: 'Skipped (createWebhook not available)',
          details: 'The createWebhook method is not available in the SDK',
        };
      }

      const result = await sdk.createWebhook({
        url: 'https://example.com/webhook/health-check',
        events: ['charge.created', 'charge.completed'],
      });
      const duration = performance.now() - start;

      // Store webhook ID for subsequent checks
      (ctx as WebhookContext).webhookId = result.id;

      return {
        name: 'Webhook Create',
        success: true,
        duration,
        message: `Created: ${result.id}`,
        details: `URL: ${result.url}, Events: ${result.events.join(', ')}`,
      };
    } catch (error) {
      const duration = performance.now() - start;
      const err = error as DripError;

      if (err.statusCode === 404 || err.statusCode === 501) {
        return {
          name: 'Webhook Create',
          success: true,
          duration,
          message: 'Skipped (endpoint not implemented)',
          details: `Status: ${err.statusCode}`,
        };
      }

      return {
        name: 'Webhook Create',
        success: false,
        duration,
        message: err.message || 'Failed to create webhook',
        details: `${err.code} (status: ${err.statusCode})`,
      };
    }
  },
};

export const webhookListCheck: Check = {
  name: 'Webhook List',
  description: 'List all webhooks',
  async run(ctx: CheckContext): Promise<CheckResult> {
    const start = performance.now();
    const client = createClient(ctx);

    try {
      const result = await client.listWebhooks();
      const duration = performance.now() - start;

      // Store the first webhook ID if available for subsequent checks
      const webhooks = result as unknown as { data?: Array<{ id: string }> } | Array<{ id: string }>;
      const webhookArray = Array.isArray(webhooks) ? webhooks : webhooks.data || [];
      if (webhookArray.length > 0 && !(ctx as WebhookContext).webhookId) {
        (ctx as WebhookContext).webhookId = webhookArray[0].id;
      }

      const count = webhookArray.length;

      return {
        name: 'Webhook List',
        success: true,
        duration,
        message: `Found ${count} webhook(s)`,
      };
    } catch (error) {
      const duration = performance.now() - start;
      const err = error as DripError;

      if (err.statusCode === 404 || err.statusCode === 501) {
        return {
          name: 'Webhook List',
          success: true,
          duration,
          message: 'Skipped (endpoint not implemented)',
          details: `Status: ${err.statusCode}`,
        };
      }

      return {
        name: 'Webhook List',
        success: false,
        duration,
        message: err.message || 'Failed to list webhooks',
        details: `${err.code} (status: ${err.statusCode})`,
      };
    }
  },
};

export const webhookGetCheck: Check = {
  name: 'Webhook Get',
  description: 'Get webhook by ID',
  async run(ctx: CheckContext): Promise<CheckResult> {
    const start = performance.now();
    const webhookId = (ctx as WebhookContext).webhookId;

    if (!webhookId) {
      return {
        name: 'Webhook Get',
        success: true,
        duration: 0,
        message: 'Skipped (no webhook ID available)',
        suggestion: 'Run Webhook Create or Webhook List first',
      };
    }

    try {
      const client = createClient(ctx);
      const sdk = (client as unknown as {
        sdk: {
          getWebhook?: (id: string) => Promise<{ id: string; url: string; events: string[] }>;
        };
      }).sdk;

      if (!sdk.getWebhook) {
        const duration = performance.now() - start;
        return {
          name: 'Webhook Get',
          success: true,
          duration,
          message: 'Skipped (getWebhook not available)',
          details: 'The getWebhook method is not available in the SDK',
        };
      }

      const result = await sdk.getWebhook(webhookId);
      const duration = performance.now() - start;

      return {
        name: 'Webhook Get',
        success: true,
        duration,
        message: `Got webhook: ${result.id}`,
        details: `URL: ${result.url}`,
      };
    } catch (error) {
      const duration = performance.now() - start;
      const err = error as DripError;

      if (err.statusCode === 404 || err.statusCode === 501) {
        return {
          name: 'Webhook Get',
          success: true,
          duration,
          message: 'Skipped (endpoint not implemented)',
          details: `Status: ${err.statusCode}`,
        };
      }

      return {
        name: 'Webhook Get',
        success: false,
        duration,
        message: err.message || 'Failed to get webhook',
        details: `${err.code} (status: ${err.statusCode})`,
      };
    }
  },
};

export const webhookTestCheck: Check = {
  name: 'Webhook Test',
  description: 'Send test webhook event',
  async run(ctx: CheckContext): Promise<CheckResult> {
    const start = performance.now();
    const webhookId = (ctx as WebhookContext).webhookId;

    if (!webhookId) {
      return {
        name: 'Webhook Test',
        success: true,
        duration: 0,
        message: 'Skipped (no webhook ID available)',
        suggestion: 'Run Webhook Create or Webhook List first',
      };
    }

    try {
      const client = createClient(ctx);
      const sdk = (client as unknown as {
        sdk: {
          testWebhook?: (id: string) => Promise<{ success: boolean; statusCode?: number }>;
        };
      }).sdk;

      if (!sdk.testWebhook) {
        const duration = performance.now() - start;
        return {
          name: 'Webhook Test',
          success: true,
          duration,
          message: 'Skipped (testWebhook not available)',
          details: 'The testWebhook method is not available in the SDK',
        };
      }

      const result = await sdk.testWebhook(webhookId);
      const duration = performance.now() - start;

      return {
        name: 'Webhook Test',
        success: true,
        duration,
        message: result.success ? 'Test event sent' : 'Test event failed',
        details: result.statusCode ? `Response status: ${result.statusCode}` : undefined,
      };
    } catch (error) {
      const duration = performance.now() - start;
      const err = error as DripError;

      if (err.statusCode === 404 || err.statusCode === 501) {
        return {
          name: 'Webhook Test',
          success: true,
          duration,
          message: 'Skipped (endpoint not implemented)',
          details: `Status: ${err.statusCode}`,
        };
      }

      return {
        name: 'Webhook Test',
        success: false,
        duration,
        message: err.message || 'Failed to test webhook',
        details: `${err.code} (status: ${err.statusCode})`,
      };
    }
  },
};

export const webhookRotateSecretCheck: Check = {
  name: 'Webhook Rotate Secret',
  description: 'Rotate webhook signing secret',
  async run(ctx: CheckContext): Promise<CheckResult> {
    const start = performance.now();
    const webhookId = (ctx as WebhookContext).webhookId;

    if (!webhookId) {
      return {
        name: 'Webhook Rotate Secret',
        success: true,
        duration: 0,
        message: 'Skipped (no webhook ID available)',
        suggestion: 'Run Webhook Create or Webhook List first',
      };
    }

    try {
      const client = createClient(ctx);
      const sdk = (client as unknown as {
        sdk: {
          rotateWebhookSecret?: (id: string) => Promise<{ secret: string }>;
        };
      }).sdk;

      if (!sdk.rotateWebhookSecret) {
        const duration = performance.now() - start;
        return {
          name: 'Webhook Rotate Secret',
          success: true,
          duration,
          message: 'Skipped (rotateWebhookSecret not available)',
          details: 'The rotateWebhookSecret method is not available in the SDK',
        };
      }

      const result = await sdk.rotateWebhookSecret(webhookId);
      const duration = performance.now() - start;

      return {
        name: 'Webhook Rotate Secret',
        success: true,
        duration,
        message: 'Secret rotated successfully',
        details: `New secret: ${result.secret.slice(0, 10)}...`,
      };
    } catch (error) {
      const duration = performance.now() - start;
      const err = error as DripError;

      if (err.statusCode === 404 || err.statusCode === 501) {
        return {
          name: 'Webhook Rotate Secret',
          success: true,
          duration,
          message: 'Skipped (endpoint not implemented)',
          details: `Status: ${err.statusCode}`,
        };
      }

      return {
        name: 'Webhook Rotate Secret',
        success: false,
        duration,
        message: err.message || 'Failed to rotate webhook secret',
        details: `${err.code} (status: ${err.statusCode})`,
      };
    }
  },
};

export const webhookDeleteCheck: Check = {
  name: 'Webhook Delete',
  description: 'Delete webhook (cleanup)',
  async run(ctx: CheckContext): Promise<CheckResult> {
    const start = performance.now();
    const webhookId = (ctx as WebhookContext).webhookId;

    if (!webhookId) {
      return {
        name: 'Webhook Delete',
        success: true,
        duration: 0,
        message: 'Skipped (no webhook ID to delete)',
      };
    }

    try {
      const client = createClient(ctx);
      const sdk = (client as unknown as {
        sdk: {
          deleteWebhook?: (id: string) => Promise<void>;
        };
      }).sdk;

      if (!sdk.deleteWebhook) {
        const duration = performance.now() - start;
        return {
          name: 'Webhook Delete',
          success: true,
          duration,
          message: 'Skipped (deleteWebhook not available)',
          details: 'The deleteWebhook method is not available in the SDK',
        };
      }

      await sdk.deleteWebhook(webhookId);
      const duration = performance.now() - start;

      // Clear the webhook ID
      delete (ctx as WebhookContext).webhookId;

      return {
        name: 'Webhook Delete',
        success: true,
        duration,
        message: `Deleted webhook: ${webhookId}`,
      };
    } catch (error) {
      const duration = performance.now() - start;
      const err = error as DripError;

      if (err.statusCode === 404 || err.statusCode === 501) {
        return {
          name: 'Webhook Delete',
          success: true,
          duration,
          message: 'Skipped (endpoint not implemented)',
          details: `Status: ${err.statusCode}`,
        };
      }

      return {
        name: 'Webhook Delete',
        success: false,
        duration,
        message: err.message || 'Failed to delete webhook',
        details: `${err.code} (status: ${err.statusCode})`,
      };
    }
  },
};

File: python/checks/webhooks_crud.py
"""Webhook CRUD operation checks."""
import uuid
from ..types import Check, CheckContext, CheckResult
from ..drip_client import create_client


async def _webhook_create_check(ctx: CheckContext) -> CheckResult:
    """Create a webhook endpoint."""
    try:
        client = create_client(ctx.api_key, ctx.api_url)

        if not hasattr(client, 'create_webhook'):
            return CheckResult(
                name="webhook_create",
                success=True,
                duration=0,
                message="Skipped (create_webhook not available)",
                details="The create_webhook method is not available in the SDK"
            )

        webhook_url = f"https://webhook.site/health-check-{uuid.uuid4().hex[:8]}"
        result = client.create_webhook(
            url=webhook_url,
            events=["charge.created", "charge.settled"]
        )

        webhook_id = getattr(result, 'id', str(result))
        secret = getattr(result, 'secret', None)

        ctx.webhook_id = webhook_id
        if secret:
            ctx.webhook_secret = secret

        return CheckResult(
            name="webhook_create",
            success=True,
            duration=0,
            message=f"Created webhook {webhook_id}",
            details=f"URL: {webhook_url}"
        )
    except Exception as e:
        error_str = str(e)
        if '404' in error_str or '501' in error_str:
            return CheckResult(
                name="webhook_create",
                success=True,
                duration=0,
                message="Skipped (endpoint not implemented)",
                details=error_str
            )
        return CheckResult(
            name="webhook_create",
            success=False,
            duration=0,
            message=f"Failed to create webhook: {e}"
        )


webhook_create_check = Check(
    name="webhook_create",
    description="Create webhook endpoint",
    run=_webhook_create_check
)


async def _webhook_list_check(ctx: CheckContext) -> CheckResult:
    """List all webhooks."""
    try:
        client = create_client(ctx.api_key, ctx.api_url)

        if not hasattr(client, 'list_webhooks'):
            return CheckResult(
                name="webhook_list",
                success=True,
                duration=0,
                message="Skipped (list_webhooks not available)",
                details="The list_webhooks method is not available in the SDK"
            )

        result = client.list_webhooks()

        if hasattr(result, 'data'):
            count = len(result.data)
        elif isinstance(result, list):
            count = len(result)
        else:
            count = 1

        return CheckResult(
            name="webhook_list",
            success=True,
            duration=0,
            message=f"Found {count} webhook(s)"
        )
    except Exception as e:
        error_str = str(e)
        if '404' in error_str or '501' in error_str:
            return CheckResult(
                name="webhook_list",
                success=True,
                duration=0,
                message="Skipped (endpoint not implemented)",
                details=error_str
            )
        return CheckResult(
            name="webhook_list",
            success=False,
            duration=0,
            message=f"Failed to list webhooks: {e}"
        )


webhook_list_check = Check(
    name="webhook_list",
    description="List all webhooks",
    run=_webhook_list_check
)


async def _webhook_get_check(ctx: CheckContext) -> CheckResult:
    """Get webhook by ID."""
    if not ctx.webhook_id:
        return CheckResult(
            name="webhook_get",
            success=True,
            duration=0,
            message="Skipped (no webhook ID)",
            details="Run webhook_create check first"
        )

    try:
        client = create_client(ctx.api_key, ctx.api_url)

        if not hasattr(client, 'get_webhook'):
            return CheckResult(
                name="webhook_get",
                success=True,
                duration=0,
                message="Skipped (get_webhook not available)",
                details="The get_webhook method is not available in the SDK"
            )

        webhook = client.get_webhook(ctx.webhook_id)
        url = getattr(webhook, 'url', 'N/A')

        return CheckResult(
            name="webhook_get",
            success=True,
            duration=0,
            message=f"Retrieved webhook {ctx.webhook_id}",
            details=f"URL: {url}"
        )
    except Exception as e:
        error_str = str(e)
        if '404' in error_str or '501' in error_str:
            return CheckResult(
                name="webhook_get",
                success=True,
                duration=0,
                message="Skipped (endpoint not implemented)",
                details=error_str
            )
        return CheckResult(
            name="webhook_get",
            success=False,
            duration=0,
            message=f"Failed to get webhook: {e}"
        )


webhook_get_check = Check(
    name="webhook_get",
    description="Get webhook by ID",
    run=_webhook_get_check
)


async def _webhook_test_check(ctx: CheckContext) -> CheckResult:
    """Send test webhook event."""
    if not ctx.webhook_id:
        return CheckResult(
            name="webhook_test",
            success=True,
            duration=0,
            message="Skipped (no webhook ID)",
            details="Run webhook_create check first"
        )

    try:
        client = create_client(ctx.api_key, ctx.api_url)

        if not hasattr(client, 'test_webhook'):
            return CheckResult(
                name="webhook_test",
                success=True,
                duration=0,
                message="Skipped (test_webhook not available)",
                details="The test_webhook method is not available in the SDK"
            )

        result = client.test_webhook(ctx.webhook_id)
        sent = getattr(result, 'sent', getattr(result, 'success', True))

        return CheckResult(
            name="webhook_test",
            success=True,
            duration=0,
            message=f"Test event sent: {sent}"
        )
    except Exception as e:
        error_str = str(e)
        if '404' in error_str or '501' in error_str:
            return CheckResult(
                name="webhook_test",
                success=True,
                duration=0,
                message="Skipped (endpoint not implemented)",
                details=error_str
            )
        return CheckResult(
            name="webhook_test",
            success=False,
            duration=0,
            message=f"Failed to test webhook: {e}"
        )


webhook_test_check = Check(
    name="webhook_test",
    description="Send test webhook event",
    run=_webhook_test_check
)


async def _webhook_rotate_secret_check(ctx: CheckContext) -> CheckResult:
    """Rotate webhook signing secret."""
    if not ctx.webhook_id:
        return CheckResult(
            name="webhook_rotate_secret",
            success=True,
            duration=0,
            message="Skipped (no webhook ID)",
            details="Run webhook_create check first"
        )

    try:
        client = create_client(ctx.api_key, ctx.api_url)

        if not hasattr(client, 'rotate_webhook_secret'):
            return CheckResult(
                name="webhook_rotate_secret",
                success=True,
                duration=0,
                message="Skipped (rotate_webhook_secret not available)",
                details="The rotate_webhook_secret method is not available in the SDK"
            )

        result = client.rotate_webhook_secret(ctx.webhook_id)
        new_secret = getattr(result, 'secret', 'rotated')

        return CheckResult(
            name="webhook_rotate_secret",
            success=True,
            duration=0,
            message="Secret rotated successfully",
            details=f"New secret: {new_secret[:10]}..." if len(str(new_secret)) > 10 else f"New secret: {new_secret}"
        )
    except Exception as e:
        error_str = str(e)
        if '404' in error_str or '501' in error_str:
            return CheckResult(
                name="webhook_rotate_secret",
                success=True,
                duration=0,
                message="Skipped (endpoint not implemented)",
                details=error_str
            )
        return CheckResult(
            name="webhook_rotate_secret",
            success=False,
            duration=0,
            message=f"Failed to rotate secret: {e}"
        )


webhook_rotate_secret_check = Check(
    name="webhook_rotate_secret",
    description="Rotate webhook signing secret",
    run=_webhook_rotate_secret_check
)


async def _webhook_delete_check(ctx: CheckContext) -> CheckResult:
    """Delete webhook."""
    if not ctx.webhook_id:
        return CheckResult(
            name="webhook_delete",
            success=True,
            duration=0,
            message="Skipped (no webhook ID)",
            details="Run webhook_create check first"
        )

    try:
        client = create_client(ctx.api_key, ctx.api_url)

        if not hasattr(client, 'delete_webhook'):
            return CheckResult(
                name="webhook_delete",
                success=True,
                duration=0,
                message="Skipped (delete_webhook not available)",
                details="The delete_webhook method is not available in the SDK"
            )

        client.delete_webhook(ctx.webhook_id)

        return CheckResult(
            name="webhook_delete",
            success=True,
            duration=0,
            message=f"Deleted webhook {ctx.webhook_id}"
        )
    except Exception as e:
        error_str = str(e)
        if '404' in error_str or '501' in error_str:
            return CheckResult(
                name="webhook_delete",
                success=True,
                duration=0,
                message="Skipped (endpoint not implemented)",
                details=error_str
            )
        return CheckResult(
            name="webhook_delete",
            success=False,
            duration=0,
            message=f"Failed to delete webhook: {e}"
        )


webhook_delete_check = Check(
    name="webhook_delete",
    description="Delete webhook",
    run=_webhook_delete_check
)
